### PR TITLE
Notebook/hierarchy flow

### PR DESCRIPTION
**Pull request by joshkyh**
## Why are these changes needed?

This notebook demonstrates how to control the conversation flow using select_speaker of the GroupChat class.

image

## Related issue number
nil

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+      "Thank you, A1. So B2, the product of x and y is 45.\n",
+      "\n",
+      "--------------------------------------------------------------------------------\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Initialization\n",
+    "agents_A = [\n",
+    "    AssistantAgent(name='A1', system_message=\"You are a team leader A1, your team consists of A2, A3. You can talk to the other team leader B1, whose team member is B2.\"),\n",
```
Comment (gagb) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:303]: Where is llm_config passed to these assistant agents?

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    "    AssistantAgent(name='A3', system_message=\"You are team member A3, You know the secret value of y but not x, y = 5. Tell others y to cooperate.\"),\n",
+    "]\n",
+    "\n",
+    "agents_B = [\n",
+    "    AssistantAgent(name='B1', system_message=\"You are a team leader B1, your team consists of B2. You can talk to the other team leader A1, whose team member is A2, A3. Use NEXT: A1 to suggest talking to A1.\"),\n",
+    "    AssistantAgent(name='B2', system_message=\"You are team member B2. Your task is to find out the value of x and y and compute the product. Once you have the answer, say out the answer and append a new line with TERMINATE.\")\n",
+    "]\n",
+    "\n",
+    "# Terminates the conversation when TERMINATE is detected.\n",
+    "user_proxy = autogen.UserProxyAgent(\n",
```
Comment (gagb) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:314]: there is a system message, execution is turned off, and human input is never. But there is no llm_config

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+    "# Create CustomGroupChat\n",
+    "group_chat = CustomGroupChat(\n",
+    "    agents=list_of_agents,  # Include all agents\n",
+    "    messages=['Everyone cooperate and help agent B2 in his task. Team A has A1, A2, A3. Team B has B1, B2. Only members of the same team can talk to one another. Only team leaders (names ending with 1) can talk amongst themselves. You must use \"NEXT: B1\" to suggest talking to B1 for example; you cannot suggest yourself or the previous speaker; You can also dont suggest anyone.'],\n",
+    "    max_round=30\n",
+    ")\n",
+    "\n",
+    "\n",
+    "# Create the manager\n",
+    "llm_config = {\"config_list\": config_list_gpt4, \"seed\": 42, \"use_cache\":False}  # Assuming config_list_gpt4 is defined elsewhere\n",
```
Comment (gagb) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:333]: What is the reason for `user_cache=False` here?

**Comment by joshkyh** [comment]
Hi @gagb, thank you for the review, I have worked on the three points you've brought up.
1. Added llm_config to all 5 assistant agents.
2. Added llm_config to 1 user proxy agent.
3. Clarfied why use_cache was set to false.

**Review by gagb — APPROVED** [review]
This PR adds an example that demonstrates how one could create custom group chat logic. It will help others create custom such examples. The example in the notebook is very cleverly designed to demonstrate information flow between the agents.

**Comment by joshkyh** [comment]
> This PR adds an example that demonstrates how one could create custom group chat logic. It will help others create custom such examples. The example in the notebook is very cleverly designed to demonstrate information flow between the agents.

Thank you for the kind words. :)

File: test/test_notebook.py [hunk 4]
```diff
@@ -83,6 +83,9 @@ def test_agentchat_MathChat(save=False):
 def test_oai_chatgpt_gpt4(save=False):
     run_notebook("oai_chatgpt_gpt4.ipynb", save=save)
 
+def test_hierarchy_flow_using_select_speaker(save=False):
```
Comment (sonichi) [inline test/test_notebook.py:86]: Could you please add a `@pytest.mark.skipif` decorator like other tests to run it only in one py version, 3.10 or 3.11?

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+      "\n",
+      "--------------------------------------------------------------------------------\n",
+      "[autogen.oai.completion: 10-31 09:38:16] {788} WARNING - Completion was provided with a config_list, but the list was empty. Adopting default OpenAI behavior, which reads from the 'model' parameter instead.\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "WARNING:autogen.oai.completion:Completion was provided with a config_list, but the list was empty. Adopting default OpenAI behavior, which reads from the 'model' parameter instead.\n"
```
Comment (sonichi) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:273]: This warning implies that the config_list is not loaded properly. Is it intentional?

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[33mA1\u001b[0m (to chat_manager):\n",
+      "\n",
+      "Yes, A2 has the value of x and A3 knows the value of y. Let me get those for you. A2, what is the value of x? And A3, what is the value of y?\n",
```
Comment (sonichi) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:302]: This is the last message. Shouldn't there be more messages in the conversation?

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+    "]\n",
+    "\n",
+    "# Terminates the conversation when TERMINATE is detected.\n",
+    "user_proxy = autogen.UserProxyAgent(\n",
+    "        name=\"User_proxy\",\n",
+    "        system_message=\"Terminator admin.\",\n",
+    "        code_execution_config=False,\n",
+    "        is_termination_msg=is_termination_msg,\n",
+    "        human_input_mode=\"NEVER\",\n",
+    "        llm_config=llm_config)\n",
```
Comment (sonichi) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:346]: Does the user proxy agent need to use llm to generate reply? If not, we shouldn't set its llm_config.

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+    "\n",
+    "list_of_agents =  agents_A + agents_B\n",
```
Comment (sonichi) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:348]:
```suggestion
    "list_of_agents = agents_A + agents_B\n",
```

**Review by sonichi — COMMENTED** [review]
Thanks for the PR. The notebook output seems incomplete, and there is some warning which needs to be checked.

**Comment by gagb** [comment]
> Thanks for the PR. The notebook output seems incomplete, and there is some warning which needs to be checked.

It does look incomplete. Tho on my local run it output messages consistent with the diagram. @joshkyh did you commit an earlier version of your nb?

**Comment by joshkyh** [comment]
Thanks both, I would work on this over the next 24 hours.

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb (hunk 6 above)
Comment (joshkyh) [reply notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:302]: I'm not sure what happened there, but anyways, I have cleared all outputs and rerun the entire script.



B2 (to chat_manager):

Find the product of x and y, the other agents know x and y.

--------------------------------------------------------------------------------
B1 (to chat_manager):

NEXT: A1
Can you or your team member provide the value of x and y?

--------------------------------------------------------------------------------
Extracted suggested_next = A1
Can you or your team member provide the value of x and y?
A1 (to chat_manager):

As A1, I don't have the values. A2, A3, do you know the values of x and y?

--------------------------------------------------------------------------------
A2 (to chat_manager):

As A2, I can confirm that the value of x is 9. However, I do not know the value of y, maybe another team member could provide it?

--------------------------------------------------------------------------------
A3 (to chat_manager):

As A3, I can confirm that the value of y is 5.

--------------------------------------------------------------------------------
A1 (to chat_manager):

Considering that the value of x is 9 and the value of y is 5, the product of x and y is 45.

--------------------------------------------------------------------------------
B1 (to chat_manager):

NEXT: B2
The product of x and y is 45.

--------------------------------------------------------------------------------
Extracted suggested_next = B2
The product of x and y is 45.
B2 (to chat_manager):

The product of x and y is 45. 
TERMINATE

--------------------------------------------------------------------------------

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb (hunk 5 above)
Comment (joshkyh) [reply notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb]: I have corrected the path and there is no more warning.
image

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb (hunk 7 above)
Comment (joshkyh) [reply notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb]: Got it, removed the llm_config from user proxy agent.

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb (hunk 8 above)
Comment (joshkyh) [reply notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb]: Apologies, amended type.

**Comment by joshkyh** [comment]
> Thanks for the PR. The notebook output seems incomplete, and there is some warning which needs to be checked.

Thanks for the review @sonichi, please see responses in-line with your comments.

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb [hunk 9]
```diff
@@ -0,0 +1,392 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a href=\"https://colab.research.google.com/github/microsoft/autogen/blob/main/notebook/agentchat_chess.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (sonichi) [inline notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb:8]: This link is outdated. :)
We can merge after changing it.
```suggestion
    "<a href=\"https://colab.research.google.com/github/microsoft/autogen/blob/main/notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```

**Review by sonichi — COMMENTED** [review]
Would you like to add a link in https://microsoft.github.io/autogen/docs/Examples/AutoGen-AgentChat and https://microsoft.github.io/autogen/docs/Use-Cases/agent_chat#diverse-applications-implemented-with-autogen ? It's OK to do it in a separate PR.

File: notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb (hunk 9 above)
Comment (joshkyh) [reply notebook/agentchat_hierarchy_flow_using_select_speaker.ipynb]: Sorry about that, I didn't realise. Please see updated link in the new commit.
image

**Comment by joshkyh** [comment]
> Would you like to add a link in https://microsoft.github.io/autogen/docs/Examples/AutoGen-AgentChat and https://microsoft.github.io/autogen/docs/Use-Cases/agent_chat#diverse-applications-implemented-with-autogen ? It's OK to do it in a separate PR.

Yes, no problem, I'll work on this in a new PR. :)

**Comment by joshkyh** [comment]
> Would you like to add a link in https://microsoft.github.io/autogen/docs/Examples/AutoGen-AgentChat and https://microsoft.github.io/autogen/docs/Use-Cases/agent_chat#diverse-applications-implemented-with-autogen ? It's OK to do it in a separate PR.

Yes OK. I'll work on this in a separate PR.

**Comment by sonichi** [comment]
There is still a code format issue: https://github.com/microsoft/autogen/actions/runs/6761693140/job/18376828027?pr=482 to fix before merging.

**Comment by joshkyh** [comment]
> There is still a code format issue: https://github.com/microsoft/autogen/actions/runs/6761693140/job/18376828027?pr=482 to fix before merging. BTW, have you tested the new pyautogen version v0.2.0b1? "api_base" has been renamed into "base_url".

Thanks for the review @sonichi.

1. I have rebuilt my .devcontainer to use the 0.2.0b1 release, and changed to "base_url".

2. I have passed pre-commit run --all-files locally before pushing this time, apologies. Still learning pre-commit. :)
image

**Comment by joshkyh** [comment]
Thanks for the patient guidance everyone!

**Pull request merged**